### ISSUE TITLE
[AXE-CORE] Heading levels should increase by one

### ISSUE THREAD

**Issue by bkjohnson** [department-of-veterans-affairs/va.gov-team#29466]
[508-defect-2](https://github.com/department-of-veterans-affairs/va.gov-team/blob/master/platform/accessibility/guidance/defect-severity-rubric.md#508-defect-2)

## User Story or Problem Statement

A CMS content editor adds content to a page using `<h3>` tags as the main non-title content heading. They also check (or don't uncheck) the box that says "Include table of contents". This results in only 1 `<h2>` being on the page which says "On this page", and this is unexpected behavior. If the box is _not_ checked, then the headings go from `<h1>` to `<h3>`, which is a problem for screen reader users.

## Details

Background issues:
- https://github.com/department-of-veterans-affairs/va.gov-team/issues/28249
- https://github.com/department-of-veterans-affairs/vets-design-system-documentation/issues/530

Here is a screenshot of part of the content editor experience:

![image](https://user-images.githubusercontent.com/2008881/131380970-86ee46fc-37f0-4f04-bc2e-2f7b53fa8172.png)

The "component" will [only be rendered if data in the content model says to render it](
https://github.com/department-of-veterans-affairs/content-build/blob/f89ea0493ef4de92fee3344ba427cbdaf9b1cb2a/src/site/facilities/health_care_region_detail_content.drupal.liquid#L21-L26).

Based on the [documentation for the "On this page" component](https://design.va.gov/components/on-this-page-jump-links#how-to-use-the-on-this-page-jump-link-component), it is being used incorrectly on this specific page, and likely others. However, the improper use is due to there being no other `<h2>` elements on the page, which is because of bad content coming from the CMS.

> The content on the page should have a minimum of two H2 section titles to use the On this page jump link. 

## Acceptance Criteria

- [ ] Daily a11y scans can tell if there is a `heading-order` issue on the page
- [ ] (optional) The presence of the "On this page" component doesn't prevent the scans from finding a `heading-order` issue

## Environment

- Any OS
- Any browser
- Production environment

## Steps to Recreate

1. Enter https://www.va.gov/pittsburgh-health-care/health-services/surgery-service/ in browser
2. Notice that there are no other headings under "On this page"
3. Inspect the page and notice that "Surgery Service" is an `<h1>`, "On this page" is an `<h2>`, and "Contact info" is an `<h3>`
4. None of the `<h3>` elements are children of the `<h2>`.

## Proposed Solution (if known)

Modify the [`axeCheck` Cypress command](https://github.com/department-of-veterans-affairs/content-build/blob/1f638dc08816f3f443e337b428f4b0ab1d3eb5ad/src/platform/testing/e2e/cypress/support/commands/axeCheck.js#L32-L65) and/or the 16 [sitemap tests where `axeCheck` is being called](https://github.com/department-of-veterans-affairs/content-build/blob/1f638dc08816f3f443e337b428f4b0ab1d3eb5ad/src/platform/site-wide/tests/sitemap/sitemap-1.a11y.spec.js#L40-L45) to:

1. Throw errors on the `heading-order` rule in the `best-practice` tag
1. (optional) `exclude` the "On this page" component from the checks

If excluding the component and its `<h2>On this page</h2>` isn't possible, then the design system team will be able to change the component to not render the `<h2>` if there are no others on the page, leading to an appropriate failure for the a11y tests.

## Screenshots or Trace Logs

With the `best-practice` tag enabled the tests are saying that the page is missing an `<h1>`, but aren't picking up on the `heading-order` issue:

https://github.com/department-of-veterans-affairs/content-build/runs/3429054614?check_suite_focus=true#step:9:3990

This may be related to the page being inside of an `<iframe>` in Cypress.

`heading-order` was getting triggered on other pages though: https://github.com/department-of-veterans-affairs/content-build/runs/3429054717?check_suite_focus=true#step:9:2506

**Comment by noahgelman** [comment]
@bkjohnson If there is a heading issue, you can pass a different heading, like an `<h2>` instead of an `<h3>` and use a CSS class that makes it look like an a header with the `<h3>` design if that's what you want. We have those classes for this read

**Comment by bkjohnson** [comment]
I'm aware of the CSS classes for styling headings as other levels. The issue here is content is being delivered from the CMS that already has this violation. At that point the content build could try to recognize this and replace `<h3>`s with `<h2>`s where appropriate, but the hope is for the CMS to deliver correct content to begin with.

**Comment by bkjohnson** [comment]
Also, several months ago I [wrote a script](https://github.com/department-of-veterans-affairs/content-build/compare/find-heading-order-violation) to create a list of files where this violation occurs. I've run it sporadically, with the most recent being today. Here are the results:

- Sept. 02 2021 - 207 files
- Sept. 08  2021 - 212 files
- Nov. 09 2021 - 409 files
- Jan. 31 2022 - 554 files

This is all content that is on va.gov. For example, https://www.va.gov/western-colorado-health-care/work-with-us/internships-and-fellowships/psychiatric-mental-health-nurse-practitioner-residency

![Missing h2 between h1 and h3](https://user-images.githubusercontent.com/2008881/151883824-b55a9be4-e0e5-4b68-a127-311c5d2112f8.png)

Without anything in place to prevent this the number of files with improper content will continue to grow.

**Comment by noahgelman** [comment]
@timwright12 Do you have an opinion on this? You've worked with axe rules recently

**Comment by timwright12** [comment]
@noahgelman @bkjohnson I would like to see us start to take some steps to identify and remediate this issue.

**Heading order related to:**

- [Success Criterion 1.3.1: Info and Relationships](https://www.w3.org/WAI/WCAG21/Understanding/info-and-relationships)
- [Success Criterion 2.4.10: Section Headings](https://www.w3.org/WAI/WCAG21/Understanding/section-headings)

If there's even a gray area here related to compliance, we should take steps to try and fix it. This seems like a pretty widespread issue within the site content; if we can work with @cmaeng and the CMS team on a remediation path I think that would be best.

@cmaeng do you think we can all chat about this one? The issue seems to be getting worse with time. cc @timcosgrove

**Comment by cmaeng** [comment]
@timwright12 yup, let's chat. i'll reach out in slack to coordinate. will want to include @timcosgrove @kevwalsh @swirtSJW

**Comment by kevwalsh** [comment]
@bkjohnson This is great, thanks so much for this data. 

Do you mind 
1. running it again and getting the latest data?
2. providing the full output from this script, so that we can analyze it for patterns and opportunities.

We have a sense about where these problems are coming from (My hunch is mostly pages built with the "VAMC detail page" content type, which is used extensively by VAMC editors for content that doesn't have a defined content model, and editors have ability to choose headings in a rich text editor). 

We'll need a "how might we" with some hypotheses about what will most quickly move the needle on these, and i'd like to start tracking metrics to see what's working. [Here's a POC of what the metrics might be](https://docs.google.com/spreadsheets/d/12dJAmyhLSkFi6mD4u9FMqnDjpLo1XSPEg6wtBFefR6g/edit#gid=0). 

<img width="931" alt="Heading_errors_data_-_Google_Sheets" src="https://user-images.githubusercontent.com/643678/156452413-92ef7982-d444-4b0c-baae-0d7751be0809.png">


In the future, we might want to create charts like this specific to each vector, like per product, per content type, or even per user, if we find that it's a few users who are creating a disproportionate number of these. 

FWIW, one hypothesis is that editors think h2 is too big and so they choose h3 for purely aesthetic reasons, without understanding the ramifications.

**Comment by timwright12** [comment]
Release tools action item ticket: https://github.com/department-of-veterans-affairs/va.gov-team/issues/37996

**Comment by bkjohnson** [comment]
@kevwalsh here's the latest results from the script:

- April 20, 2022 - 629 files

And here's the full output:

[heading_order_violations_2022-04-20.txt](https://github.com/department-of-veterans-affairs/va.gov-team/files/8525673/heading_order_violations.txt)

I'm not the only one that can run this - `yarn list-heading-order-violations` on the `find-heading-order-violation` branch will create the list of files.

**Comment by davidconlon** [comment]
@bkjohnson - might need a little help understanding two of the entries and where to route them:
- /session-expired/index.html - Is this the Auth team? or is it Unauth. Not sure who owns this.
- /dst/index.html - I'm not seeing this come up on the web.


FYSA - I've routed these items to the content team for manual remediation:
/health-care/health-needs-conditions/health-issues-related-to-service-era/index.html
/housing-assistance/home-loans/index.html
/careers-employment/vocational-rehabilitation/programs/index.html

The others are part of the VAMC product that we will be discussing shortly of a path forward for both educations and (hopeful) some sort of automated remediation/prevention.

**Comment by bkjohnson** [comment]
I'm also not sure who owns the "session-expired" page, but I would guess the auth team since they [redirect to it](https://github.com/department-of-veterans-affairs/vets-website/blob/30dfd600ee6c0fd497cbd94ef55eae0fdee3ab7c/src/applications/auth/containers/AuthApp.jsx#L26).

---

https://staging.va.gov/dst shows the "Veterans Community Care Program Eligibility Tool":

![DST page](https://user-images.githubusercontent.com/2008881/165566070-e9e69422-9fd4-4b6c-876a-6f29a7de70d3.png)

I don't really know anything about this or why "dst" is used, but it wasn't obvious to me why this was in the list. I looked at the markup and saw why the script was flagging it:

![image](https://user-images.githubusercontent.com/2008881/165567454-6309127d-3d40-4935-8bc4-687ceafea2ac.png)

**Comment by davidconlon** [comment]
Do we know which team owns that codebase for DST?
Looking at the content I'm wondering if this is old and crufty, never launched, and needs to be removed or if this is something new.
I tried the form and it does nothing.

**Comment by davidconlon** [comment]
@bkjohnson @mchelen 
Update: We're working through a plan for the content that is in the CMS, but the DST page needs to be routed/remediated by another team as it is not in the CMS.

**Comment by davidconlon** [comment]
[Heading-levels-sliced-and-diced.xlsx](https://github.com/department-of-veterans-affairs/va.gov-team/files/8575686/Heading-levels-sliced-and-diced.xlsx)

Attaching my slice and dice for different ways to solve this problem.

**Comment by bkjohnson** [comment]
The DST page is coming from [the `vagov-content` repo](https://github.com/department-of-veterans-affairs/vagov-content/blob/40076427daabb650b6e47f528ac1ec9bd1810412/pages/dst/index.md) - so it looks like something that's old and can be removed.
